dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,446 @@
+                * bracket : the Hyperband bracket number. It runs several
+                  brackets to balance the explore/exploit tradeoff (should you
+                  train many models for few iterations or few models for many
+                  iterations?)
+                * bracket_iter : The iteration for this bracket. At each time
+                  step, it "kills" 1 - 1/eta of the models.
+                * partial_fit_iters: the number of times partial_fit will be
+                  called for per model in this iteration in this bracket
+                * num_models : the number of models that will be fit for
+                  ``partial_fit_iters`` iterations
+        """
+
+        import pandas as pd
+        X = da.from_array(np.random.rand(2, 2), 2)
+        y = da.from_array(np.random.rand(2, 2), 2)
+        self.fit(X, y, dry_run=True)
+        df = pd.DataFrame(self.history)
+
+        brackets = df[['bracket', 'bracket_iter', 'partial_fit_iters', 'num_models']]
+        values = {(s, i, iter, n) for _, (s, i, iter, n) in brackets.iterrows()}
+        values = sorted(values)
+        brackets = [{'bracket': v[0], 'bracket_iter': v[1],
+                     'partial_fit_iters': v[2], 'num_models': v[3]}
+                    for v in values]
+
+        return {'num_models': len(df.model_id.unique()),
+                'num_partial_fit_calls': df.partial_fit_iters.sum(),
+                'num_cv_splits': 1,
+                'brackets': brackets}
+
+
+def _LDtoDL(ld):
+    """list of dicts to dict of lists"""
+    keys = set(sum([list(d.keys()) for d in ld], []))
+    values = {k: [d.get(k, None) for d in ld] for k in keys}
+    return values
+
+
+def _get_cv_results(params=None, val_scores=None, times=None):
+    assert all([isinstance(x, dict) for x in [params, val_scores, times]])
+    assert set(params.keys()) == set(val_scores.keys()) == set(times.keys())
+
+    cv_results = []
+    best = {'index': None, 'score': -np.inf}
+    for i, k in enumerate(val_scores.keys()):
+        result = {'params': params[k], 'test_score': val_scores[k], **times[k]}
+        result.update({'param_' + param: v for param, v in params[k].items()})
+        cv_results += [result]
+
+        if result['test_score'] > best['score']:
+            best = {'index': i, 'score': result['test_score']}
+
+    cv_results = _LDtoDL(cv_results)
+    return cv_results, best['index']
+
+
+def _get_best_model(val_scores, models):
+    scores = {k: val_scores[k] for k in models}
+    best_id = max(scores, key=scores.get)
+    return models[best_id]
+
+
+def _get_client():
+    import distributed
+    try:
+        return distributed.get_client()
+    except ValueError:
+        raise ValueError('No global distributed client found with '
+                         'distributed.get_client. To resolve this error, '
+                         'have an distributed client: '
+                         'https://distributed.readthedocs.io/en/latest/client.html')
```

Comment:
I'm curious, why this function?

Reply:
I wanted to raise a clear error message. I've refactored to provide the text as a warning then raise `distributed.get_client`'s exception.